### DISCUSSION TITLE
Can L2ARCH be in a RAID1 Mirrored Device Pair Partition?

### DISCUSSION THREAD

**Discussion by spflanze** [openzfs/zfs#13556] [General]
Below is a  snippet from a command sequence that results in the error message shown in the last line: 
```
# To find which long alias in the /dev/disk/by-id/ directory is associated with which
# logical name, use the command:
lshw -C disk

# logical name: /dev/nvme0n1
SSD1=/dev/disk/by-id/nvme-eui.002538ba11517132
# logical name: /dev/nvme1n1
SSD2=/dev/disk/by-id/nvme-eui.002538ba11505423
#logical name: /dev/sda
HDD1=/dev/disk/by-id/ata-HGST_HUH721212ALE600_D5HBTJYF
#logical name: /dev/sdb
HDD2=/dev/disk/by-id/ata-HGST_HUH721212ALE600_D5HBRWVF

# Create SLOG Write Cache Partition
sgdisk     -n6:0:+10G -t6:FD00  $SSD1
sgdisk     -n6:0:+10G -t6:FD00  $SSD2

# Create L2ARC Read Cache Partition
sgdisk     -n7:0:+20G -t7:FD00  $SSD1
sgdisk     -n7:0:+20G -t7:FD00  $SSD2

# Create the home pool (hpool):
# ZFS native encryption::
zpool create \
    -o ashift=12 \
    -o autotrim=on \
    -O encryption=on -O keylocation=prompt -O keyformat=passphrase \
    -O acltype=posixacl -O xattr=sa -O dnodesize=auto \
    -O compression=lz4 \
    -O normalization=formD \
    -O relatime=on \
    -O canmount=off -O mountpoint=/ -R /mnt \
   hpool -f mirror ${HDD1}-part1  ${HDD2}-part1
   
   
   zpool add hpool -f log mirror ${SSD1}-part6 ${SSD2}-part6
   zpool add hpool -f cache mirror ${SSD1}-part7 ${SSD2}-part7
  
   invalid vdev specification: cache requires at least 1 devices
```
Why does the cache parameter not recognize the device `"mirror ${SSD1}-part7  ${SSD2}-part7"` ?

**Comment by rincebrain** [comment]
Per `zpoolconcepts(8)`:
`A cache device cannot be configured as a mirror or raidz group.`

I suppose conceptually that made more sense before persistent L2ARC went in.

**Comment by spflanze** [comment]
Thank you for your reply. So I have to give up on the idea of the cache being mirrored. So I tried to use the two partitions separately to double the available storage space for the L2ARCH. I got another error:
```
$ zpool add hpool -f cache ${SSD1}-part7 ${SSD2}-part7
invalid option 'p'
usage:
	add [-fgLnP] [-o property=value] <pool> <vdev> ...
```
I do not see that I am using any 'p' option in this command. What went wrong here?

**Reply by rincebrain** [reply]
I would assume that either SSD1 or SSD2 didn't expand, so you ended up with `zpool add hpool -f cache -part7 somedisk-part7` or the like.

You also shouldn't default to using -f, that may have unexpected outcomes sometimes if you override, say, "this disk has a real filesystem on it".

**Comment by spflanze** [comment]
You are right about SSD1. and SSD2, not expanding. I verified they are not with the echo command. All that happened between when these were expanding earlier, and now, is that I had shut down the main computer I was issuing these commands from by means of SSH. The target machine these commands were executed on by means of SSH was left running. Since these values are set in the target machine, I do not know why SSD1, and SSD2, values are lost.

I executed commands to set again the values for SSD1, and SSD2, by means of SSH while logged in as root, and still they do not expand. I do not know why. I am going to shut both computers down now, and start again from beginning Ubuntu installation in the morning.

These are the directions I have been mostly following:
https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2022.04%20Root%20on%20ZFS.html

The log, and cache, properties, as well as the hpool cache in the HDD mirrored pair is my customized addition. The Ubuntu installation is being installed in the nvme SSD pairs.

I much appreciate your help :)

**Reply by rincebrain** [reply]
Shell environment variables don't persist beyond the lifetime of the shell, so if you were logged in over SSH and logged out, they would no longer be set if you logged in again, unless you reconfigured the system to set them on login.

Ditto if you were using, say, `sudo` for some of these commands - it won't necessarily pass through your environment variables from not root, there's a flag for that (`-E`).

The log and cache devices you're adding are not properties.
